File: src/aws/apps/underthehood/ui/GuiCreation.java
/*******************************************************************************
 * Copyright 2010 Alexandros Schillings
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *   http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 ******************************************************************************/
package aws.apps.underthehood.ui;

import android.content.Context;
import android.view.Gravity;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.TableRow;
import android.widget.TextView;
import android.widget.Toast;
import aws.apps.underthehood.R;
import aws.apps.underthehood.util.UsefulBits;

public class GuiCreation {
	private final Context mContext;
	private final UsefulBits uB;
	private final LayoutInflater inflater;
	private View.OnClickListener textViewCopyClickListener;
	
	public GuiCreation(Context context) {
		super();
		this.mContext = context;
		uB = new UsefulBits(mContext);
		inflater = (LayoutInflater)mContext.getSystemService(Context.LAYOUT_INFLATER_SERVICE);
		
		textViewCopyClickListener = new View.OnClickListener() {
			@SuppressWarnings("deprecation")
			public void onClick(View v) {
				TextView t = (TextView) v;
				String text = t.getText() + "";

				if (text.length() > 0) {
					String msgtext = "";
					if (text.length()>150) {
						msgtext = text.substring(0, 150) + "...";
					} else {
						msgtext = text;
					}
					String message = "'" + msgtext + "' " + mContext.getString(R.string.text_copied);
					uB.showToast(message, Toast.LENGTH_SHORT, Gravity.TOP,0,0);

					android.text.ClipboardManager ClipMan = (android.text.ClipboardManager) mContext.getSystemService(Context.CLIPBOARD_SERVICE);
					ClipMan.setText(text);
				}
			}
		};
	}
	
	public TableRow createDataRow(String text){
		TableRow tr = (TableRow) inflater.inflate(R.layout.table_row_data, null);

		TextView text1 = (TextView) tr.findViewById(R.id.text);

		text1.setClickable(true);
		text1.setOnClickListener(textViewCopyClickListener);
		text1.setText(text);
		
		return tr;
	}
	
	public TableRow createSeperatorRow(String text){
		TableRow tr = (TableRow) inflater.inflate(R.layout.table_row_section, null);

		TextView text1 = (TextView) tr.findViewById(R.id.text);
		text1.setText(text);
		
		return tr;
	}

	public TableRow createTitleRow(String text){
		TableRow tr = (TableRow) inflater.inflate(R.layout.table_row_title, null);

		TextView text1 = (TextView) tr.findViewById(R.id.text);
		text1.setText(text);
		
		return tr;
	}
	
	public View getScrollableTable(){
		return inflater.inflate(R.layout.page_scrollable_table, null);
	}
	
	public View getScrollableTextView(){
		return inflater.inflate(R.layout.page_scrollable_textview, null);
	}
}

File: app/src/main/java/aws/apps/underthehood/ui/views/GuiCreation.kt
/*******************************************************************************
 * Copyright 2010 Alexandros Schillings
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package aws.apps.underthehood.ui.views

import android.content.ClipData
import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.widget.TableRow
import android.widget.TextView
import androidx.annotation.StringRes
import aws.apps.underthehood.R

class GuiCreation(private val mContext: Context) {
    private val userNotify: UserNotify = UserNotify(mContext.applicationContext)
    private val inflater: LayoutInflater = mContext.getSystemService(Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater
    private val textViewCopyClickListener = CopyOnClickListener(userNotify)

    fun createDataRow(text: String?): TableRow {
        val tr = inflater.inflate(R.layout.table_row_data, null) as TableRow
        val textview = tr.findViewById<TextView>(R.id.text)
        textview.isClickable = true
        textview.setOnClickListener(textViewCopyClickListener)
        textview.text = text
        return tr
    }

    fun createSeperatorRow(text: String): TableRow {
        val tr = inflater.inflate(R.layout.table_row_section, null) as TableRow
        val textView = tr.findViewById<TextView>(R.id.text)
        textView.text = text
        return tr
    }

    fun createTitleRow(text: String): TableRow {
        val tr = inflater.inflate(R.layout.table_row_title, null) as TableRow
        val textView = tr.findViewById<TextView>(R.id.text)
        textView.text = text
        return tr
    }

    fun createScrollableTable(): View {
        return inflater.inflate(R.layout.page_scrollable_table, null)
    }

    fun createScrollableTextView(): View {
        return inflater.inflate(R.layout.page_scrollable_textview, null)
    }

    fun getString(@StringRes resId: Int): String {
        return mContext.getString(resId)
    }

    private class CopyOnClickListener constructor(private val userNotify: UserNotify) : View.OnClickListener {
        override fun onClick(view: View) {
            val t = view as TextView
            val text = t.text.toString()

            if (text.isEmpty()) {
                return
            }

            val context = view.getContext()
            val msgtext = getMessageText(text)

            val message = context.getString(R.string.template_text_copied, msgtext)

            val clipboardManager = context.getSystemService(Context.CLIPBOARD_SERVICE) as android.content.ClipboardManager
            val clipData = ClipData.newPlainText("Copied data", text)
            clipboardManager.setPrimaryClip(clipData)

            userNotify.notify(message)
        }

        private fun getMessageText(original: String): String {
            return if (original.length > 150) {
                original.substring(0, 150).trim { it <= ' ' } + "..."
            } else {
                original
            }
        }
    }
}